File: src/blockchat_api/src/blockchain_module/wallet.rs
use crate::blockchain_module::transaction::Transaction; 
use rand::thread_rng;
use rsa::{RsaPrivateKey, RsaPublicKey};
use rsa::pss::{Pss, BlindedSigningKey, VerifyingKey, Signature};
use rsa::signature::{Keypair,RandomizedSigner, SignatureEncoding, Verifier};
use sha2::{Digest, Sha256};
use rand::rngs::ThreadRng;

#[derive(Debug)]
pub struct Wallet {
    pub public_key: VerifyingKey<Sha256>,
    pub private_key: BlindedSigningKey<Sha256>,
    pub transactions: Vec<Transaction>,
    pub balance: i32,
    pub node_id: i32,
    pub nonce: i32,
    pub rng: ThreadRng
}

impl Wallet {
    
    // Constructor function for struct Wallet
    pub fn _new() -> Self {
        let mut my_rng = thread_rng();

        let bits = 2048;
        let private_key = RsaPrivateKey::new(&mut my_rng, bits).expect("failed to generate a key");
        let signing_key = BlindedSigningKey::<Sha256>::new(private_key);
        let verifying_key = signing_key.verifying_key();

        Wallet {
            public_key: verifying_key,
            private_key: signing_key,
            transactions: Vec::new(),
            balance: 0,
            node_id: 0,
            nonce: 0,
            rng: my_rng
        }
    }

    // Add Transaction to Wallet
    pub fn _add_transaction(&mut self,mut _new_transaction: Transaction) -> i32 {
        self.nonce += 1;
        _new_transaction.nonce = self.nonce;
        self.transactions.push(_new_transaction);
        return self.nonce;
    }
    
    // Sign the transaction
    pub fn sign_transaction(&mut self, transaction: &Transaction) -> Option<Signature> {
        return Some(self.private_key.sign_with_rng(&mut self.rng, &transaction.data.as_bytes()));
    }

    // Verify a transaction
    pub fn verify_transaction(&self, transaction: &Transaction) -> bool {
        if let Some(signature) = &transaction.signature {
            let message = transaction.data.as_bytes();
            println!("{:?}",&message);
            self.public_key.verify(&message, &signature).expect("Could not verify transaction");
            return true;
        } else {
            false
        }
    }


}

File: legacy/API/blockchain/wallet.go
package blockchain

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"fmt"
)

type Wallet struct {
	PublicKey    *rsa.PublicKey
	PrivateKey   *rsa.PrivateKey
	Balance      int
	NodeID       int
	Nonce        int
}

func NewWallet() *Wallet {
	bits := 2048
	privateKey, err := rsa.GenerateKey(rand.Reader, bits)
	if err != nil {
		panic("failed to generate private key")
	}
	publicKey := &privateKey.PublicKey

	return &Wallet{
		PublicKey:    publicKey,
		PrivateKey:   privateKey,
		Balance:      0,
		NodeID:       0,
		Nonce:        0,
	}
}

// Add a transaction to the wallet
func (w *Wallet) AddTransaction() int {
	w.Nonce++
	return w.Nonce
}

// Sign a transaction (sender)
func (w *Wallet) SignTransaction(transaction *Transaction) ([]byte,error) {
	hashed := sha256.Sum256([]byte(transaction.Data))
	signature, err := rsa.SignPSS(rand.Reader, w.PrivateKey, crypto.SHA256, hashed[:], nil)
	if err != nil {
        fmt.Println("Could not sign signature");
		return nil,err
	}
	return signature,nil
}

// Verify a transaction (receiver)
func (w* Wallet) VerifyTransaction(Data string, Signature []byte, SenderAddress *rsa.PublicKey) (bool,error) {
	hashed := sha256.Sum256([]byte(Data))
	err := rsa.VerifyPSS(SenderAddress, crypto.SHA256, hashed[:], Signature, nil)
    return true,err
}

// Deduct money from the wallet
func (w *Wallet) DeductMoney(Amount int) bool {
    if w.Balance - Amount >= 0 {
        w.Balance -= Amount;
        return true
    }

    return false
}

// Add money to the wallet
func (w *Wallet) AddMoney(Amount int) {
    w.Balance += Amount;
}
